[#1058] Avoid concurrent cache updates when downloading modules on multiple threads

Magisus opened the pull request:
Supersedes #1080

justinstoller committed:
[#1058] Serially deploy modules that share a cachedir

When we raised the default threadpool for deploying modules above 1, we
began to see issues when caching a remote locally if that remote/cache
was shared by multiple modules within the Puppetfile.

To resolve this, when we create the module we group it with any other
module that may share its cachedir. We then pull modules by cachedir off
of the queue and serially process those.

Magisus committed:
[#1058] Add accessor for a module's cachedir

Magisus committed:
[#1038] Increase default pool_size to 4

This commit changes the default number of threads used when downloading
modules from 1 to 4.

Magisus committed:
(maint) Update changelog

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -5,8 +5,13 @@ Unreleased
+
+- Ensure that modules that share a cachedir download serially, to avoid cache corruption. [#1058](https://github.com/puppetlabs/r10k/issues/1058)
 
 ----
```
while you're in here, would you mind removing this?

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -5,8 +5,13 @@ Unreleased
+(Thanks to [mhumpula](https://github.com/mhumpula) for the [feature](https://github.com/puppetlabs/r10k/pull/823).)
+- Increase default pool_size to 4, allowing modules to be downloaded on 4 threads concurrently. [#1038](https://github.com/puppetlabs/r10k/issues/1038)
+
+## Bug Fix
```
Looks like we haven't been categorizing our changes since 3.3.3.
I know the `## Changes` above was introduced by a community PR.. 
Do we want to go back to categorizing like this? If so, do we want to fix the last few versions or nah?

Magisus replied:
No let's not.

mwaggett replied:
Let's not fix old versions or let's not re-start doing it this way at all?

Magisus replied:
Not re-start doing it.

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -5,8 +5,13 @@ Unreleased
 ----------
```
oh yeah, and would you mind putting this back to just `----`?

Magisus replied:
When rendered, this makes a line just like the shorter version. When unrendered, I kinda like that it's the same length as the word 🤷‍♀️ But I can change it if you want.

mwaggett replied:
oh interesting. then yeah, sure, we can keep it like this.

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -3,12 +3,11 @@ CHANGELOG
-
-----
+- (#823) Add filter_command configuration option for git repositories
+(Thanks to [mhumpula](https://github.com/mhumpula) for the [feature](https://github.com/puppetlabs/r10k/pull/823).)
```
(sorry I keep throwing this at your PR even though they're unrelated to your work, but) do we want to fix this so it better-matches how we've been formatting things. We seem to only mention tickets out front and we put issues and PRs on the end. ¯\_(ツ)_/¯

Magisus replied:
If we're going to be prescriptive about this, seems only fair that the format be explicitly stated somewhere.

mwaggett replied:
fair. I can add something to our PR template that has the changelog reminder.

mwaggett replied:
https://github.com/puppetlabs/r10k/pull/1082

Magisus committed:
[#1058] Add acceptance test for repeated git remote

This commit adds a beaker acceptance test verifying that it is possible
to use the same remote for several non-module objects. This case
previously broke when the `pool_size` feature was introduced, because
there was a race condition when creating the git cache in shellgit mode,
when the same remote was used in two objects being set up concurrently.

mwaggett commented on integration/tests/git_source/git_source_repeated_remote.rb:
```diff
@@ -0,0 +1,65 @@
+r10k_fqp = get_r10k_fqp(master)
+git_environments_path = '/root/environments'
+git_repo_path = '/git_repos'
+git_repo_name = 'environments'
```
this is.. confusing.. can we call this something else, like 'control-repo' or something?

mwaggett replied:
or is this just so that we don't have to configure (what file-sync calls) a submodules dir?

Magisus replied:
This corresponds to the control repo, hosted locally instead of somewhere like Github. We could call it something else, but this is the name used throughout in these tests, so I'm a little reluctant to change it just here.

mwaggett replied:
right. I'm fine leaving it if that's what we do elsewhere, I feel like it just conflates things in a way that is kind of confusing. but I assume that's it's being used so that paths look the way we expect them to, with an `environments` directory.

Magisus replied:
The name of this path shouldn't have any bearing on where things get placed, since r10k is just treating this directory/git repo like a remote. It's likely called `environments` because ultimately a control repo is just the collection of environments, each of its branches corresponding to a different one. Likely this helped the original test writer feel less confused about what was happening, even if we've gotten super used to other names (like "control repo") now.

mwaggett replied:
hmm. then I like it even less lol. but it's fine, we can leave it for consistency.

Magisus replied:
The `env_path` var above is what holds the location that content will be deployed to, assuming it doesn't have its `install_dir` overridden.

mwaggett replied:
ahh okay. so then for our purposes, we don't need to set `install_dir` at all, right?

Magisus replied:
Right. And I'm changing that, because I'm updating this to install the same module under two different names, but both in the regular modulepath.

mwaggett commented on integration/tests/git_source/git_source_repeated_remote.rb:
```diff
@@ -0,0 +1,65 @@
+  :branch => 'production'
+
+mod 'non_module_object_2',
+  :install_path => '',
```
what is the behavior when `:install_path` is an empty string?
I'm assuming `:install_path` is where to install the module and having a different path than the one for `non_module_object_1` is how we ensure these are treated as separate modules?
Could they not just go in `./mod1` and `./mod2` or something?

Magisus replied:
These are specifically not modules. 

And yeah I imagine the install path could be anything. This test is copied almost exactly from the other one that failed, minus the assertions for purging. I didn't dig much into the choices made there (the commits for these test are notoriously bare-bones, and their tickets tend to be in a different tracking system than JIRA, so the spelunking is difficult), but I can.

mwaggett replied:
I don't mind you leaving it, if that's the existing pattern we've established. just trying to understand how all things things work.

if they're not modules, what are they?

Magisus replied:
Also, `install_path` appears to be relative to the location of the Puppetfile. So this is equivalent to the `./` in the previous example, which was likely intentional in the original test, which was verifying that things at the top level (i.e. not in the module dir of an environment) would get purged.

Magisus replied:
In this specific test, they're whole control repos. But according to the docs (linked in the other comment) they're most commonly hiera data in a real install.

mwaggett replied:
so would 'content that isn't a forge module' be more accurate? like, this could be anything that lives in a git repo, right? even what we call a "module"? sorry, this is just background stuff that I'm trying to understand.

Magisus replied:
I'm going to change this to not muddy the waters in this test. But yeah, in this case, it's something that isn't structured like a Puppet module, which was a specific feature that got added. Previously, things had to look like modules for r10k to deploy them (I gather from skimming the ticket), whether they came from the Forge or from git. But now, the thing on the other end of a git source can be something else besides a Puppet module if the user wants.

mwaggett commented on integration/tests/git_source/git_source_repeated_remote.rb:
```diff
@@ -0,0 +1,65 @@
+CONF
+
+puppetfile = <<-EOS
+mod 'non_module_object_1',
```
what does it mean for this to be a 'non module object'?

Magisus replied:
The probably don't have to be for this test to work, I might change that.

But some useful links on that topic: https://tickets.puppetlabs.com/browse/CODEMGMT-649, https://puppet.com/docs/pe/2018.1/puppetfile.html#declare_git_repositories_in_the_puppetfile

mwaggett commented on integration/tests/git_source/git_source_repeated_remote.rb:
```diff
@@ -0,0 +1,68 @@
+git_repo_path = '/git_repos'
+git_repo_name = 'environments'
+git_control_remote = File.join(git_repo_path, "#{git_repo_name}.git")
+code_dir = '/etc/puppetlabs/code/environments/production'
```
Should the `env_path` be incorporated into this somehow? Is it `"#{env_path}/production"` or `"/etc/puppetlabs/code/#{env_path}/production"`? Looks like the former, based on https://puppet.com/docs/puppet/5.5/environments_creating.html#environmentpath .

Magisus replied:
The `env_path` is the absolute path, so `/etc/puppetlabs/code/environments`. But yeah it can be incorporated.

mwaggett replied:
cool, yeah, I think if we include `env_path`, it's clearer that the `git_repo_name` has no relationship to that path.

Magisus replied:
updated

mwaggett commented on integration/tests/git_source/git_source_repeated_remote.rb:
```diff
@@ -0,0 +1,68 @@
+
+mod 'test_apache',
+  :git => 'git://github.com/puppetlabs/puppetlabs-apache.git',
+  :branch => 'master'
```
Would it be more realistic for this to be the 'test' branch?

Is it worth testing both ^the more realistic case and a case where the sources are literally identical? Or is that kind of redundant/unnecessary?

Magisus replied:
That module only has a `master` and a `release` branch in its repo, and I'm not sure if release is permanent. I need to make sure it's something that this test is unlikely to fail on.

Magisus replied:
Users in the r10k community were unable to come up with a _real_ use case that would trigger this bug. So no matter what we do here, it's not likely to match up to any known real workflow.

mwaggett replied:
gotcha, okay 👍

Magisus replied:
I could find a different module that has more branches, or create a fixture, but we use `apache` elsewhere in these tests so I went with it.

mwaggett replied:
if we don't think the two cases are different, I'm fine leaving it

dhollinger commented on lib/r10k/puppetfile.rb:
```diff
@@ -212,15 +221,22 @@ def concurrent_accept(visitor, pool_size)
     Queue.new.tap do |queue|
       visitor.visit(:puppetfile, self) do
-        modules.each { |mod| queue << mod }
+        modules_by_cachedir = modules_by_vcs_cachedir.clone
```
Am I missing something here or should `modules_by_vcs_cachedir` be an instance variable or is this appropriate since it's available as an `attr_reader`?

Magisus replied:
Correct, this calls the `attr_reader` method for which the default implementation is to simply return the instance variable of the same name. We added the reader so we could retrieve that value in the tests. This could just as easily be `@modules_by_vcs_cachedir.clone`, since we're in the class where the variable is definied, but as I understand it, it's generally good practice to use the reader method if one is provided. Makes it easier to override with a custom getter later if you end up needing to do something special when retrieving the value.

dhollinger reviewed (commented):
Just had one question, otherwise looks good to me
